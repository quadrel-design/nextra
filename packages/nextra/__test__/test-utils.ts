import prettier from 'prettier'

export function clean(content: any): Promise<string> {
  const cleanedContent = content
    .slice(content.indexOf('\n'), content.lastIndexOf('function MDXContent'))
    .trim()

  return prettier.format(cleanedContent, { parser: 'typescript' })
}
