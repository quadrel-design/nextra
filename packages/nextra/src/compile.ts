import { createProcessor, ProcessorOptions } from '@mdx-js/mdx'
import { Processor } from '@mdx-js/mdx/lib/core'
import remarkGfm from 'remark-gfm'
import rehypePrettyCode from 'rehype-pretty-code'
import { remarkStaticImage } from './mdx-plugins/static-image'
import remarkHandler, { HeadingMeta } from './mdx-plugins/remark'
import { LoaderOptions } from './types'
import structurize from './mdx-plugins/structurize'
import { parseMeta, attachMeta } from './mdx-plugins/rehype-handler'
import theme from './theme.json'

const createCompiler = (mdxOptions: ProcessorOptions): Processor => {
  const compiler = createProcessor(mdxOptions)
  compiler.data('headingMeta', {
    hasH1: false,
    headings: []
  })
  return compiler
}

const rehypePrettyCodeOptions = {
  theme,
  onVisitLine(node) {
    // Prevent lines from collapsing in `display: grid` mode, and
    // allow empty lines to be copy/pasted
    if (node.children.length === 0) {
      node.children = [{ type: 'text', value: ' ' }]
    }
  },
  onVisitHighlightedLine(node) {
    node.properties.className.push('highlighted')
  },
  onVisitHighlightedWord(node) {
    node.properties.className = ['highlighted']
  }
}

type NextraOptions = {
  unstable_staticImage: boolean
  unstable_flexsearch: boolean | { codeblocks: boolean }
}

export async function compileMdx(
  source: string,
  mdxOptions: LoaderOptions['mdxOptions'] = {},
  nextraOptions: NextraOptions,
  resourcePath: string
) {
  let structurizedData = {}
  const compiler = createCompiler({
    jsx: true,
    providerImportSource: '@mdx-js/react',
    remarkPlugins: [
      ...(mdxOptions.remarkPlugins || []),
      remarkGfm,
      remarkHandler,
      ...(nextraOptions.unstable_staticImage ? [remarkStaticImage] : []),
      ...(nextraOptions.unstable_flexsearch
        ? [structurize(structurizedData, nextraOptions.unstable_flexsearch)]
        : [])
    ].filter(Boolean),
    // @ts-ignore
    rehypePlugins: [
      ...(mdxOptions.rehypePlugins || []),
      parseMeta,
      [rehypePrettyCode, rehypePrettyCodeOptions],
      attachMeta
    ].filter(Boolean)
  })
  try {
    const result = await compiler.process(source)
    return {
      result: String(result),
      ...(compiler.data('headingMeta') as HeadingMeta),
      structurizedData
    }
  } catch (err) {
    console.error(`\nError compiling ${resourcePath}\n${err}\n`)
    throw err
  }
}
