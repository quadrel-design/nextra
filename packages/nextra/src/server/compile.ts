import path from 'node:path'
import type { ProcessorOptions } from '@mdx-js/mdx'
import { createProcessor } from '@mdx-js/mdx'
import type { Processor } from '@mdx-js/mdx/lib/core'
import { remarkMermaid } from '@theguild/remark-mermaid'
import { remarkNpm2Yarn } from '@theguild/remark-npm2yarn'
import rehypeKatex from 'rehype-katex'
import type { Options as RehypePrettyCodeOptions } from 'rehype-pretty-code'
import rehypePrettyCode from 'rehype-pretty-code'
import rehypeRaw from 'rehype-raw'
import remarkFrontmatter from 'remark-frontmatter'
import remarkGfm from 'remark-gfm'
import remarkMath from 'remark-math'
import remarkReadingTime from 'remark-reading-time'
import { bundledLanguages, getHighlighter } from 'shiki'
import type { Pluggable } from 'unified'
import type {
  FrontMatter,
  LoaderOptions,
  PageOpts,
  ReadingTime,
  StructurizedData
} from '../types'
import {
  CODE_BLOCK_FILENAME_REGEX,
  CWD,
  DEFAULT_LOCALE,
  ERROR_ROUTES,
  MARKDOWN_URL_EXTENSION_REGEX
} from './constants.js'
import {
  attachMeta,
  parseMeta,
  remarkCustomHeadingId,
  remarkHeadings,
  remarkLinkRewrite,
  remarkMdxDisableExplicitJsx,
  remarkMdxFrontMatter,
  remarkRemoveImports,
  remarkStaticImage,
  remarkStructurize
} from './mdx-plugins/index.js'
import { truthy } from './utils.js'

const DEFAULT_REHYPE_PRETTY_CODE_OPTIONS: RehypePrettyCodeOptions = {
  keepBackground: false,
  grid: false,
  onVisitLine(node) {
    // Prevent lines from collapsing in `display: grid` mode, and
    // allow empty lines to be copy/pasted
    if (node.children.length === 0) {
      node.children.push({ type: 'text', value: ' ' })
    }
    delete node.properties['data-line']
  },
  filterMetaString: meta => meta.replace(CODE_BLOCK_FILENAME_REGEX, ''),
  async getHighlighter(_opts) {
    const DEFAULT_OPTS = {
      themes: {
        light: 'github-light',
        dark: 'github-dark'
      },
      defaultColor: false
    } as const

    const highlighter = await getHighlighter({
      themes: Object.values(DEFAULT_OPTS.themes),
      langs: Object.keys(bundledLanguages)
    })

    const originalCodeToHtml = highlighter.codeToHtml

    return Object.assign(highlighter, {
      codeToHtml(code: string, lang: string) {
        return originalCodeToHtml(code, { lang, ...DEFAULT_OPTS })
      },
      ansiToHtml(code: string) {
        return this.codeToHtml(code, 'ansi')
      }
    })
  }
}

const cachedCompilerForFormat: Record<
  Exclude<ProcessorOptions['format'], undefined | null>,
  Processor
> = Object.create(null)

type MdxOptions = LoaderOptions['mdxOptions'] &
  Pick<ProcessorOptions, 'jsx' | 'outputFormat'>

// @ts-expect-error -- Without bind is unable to use `remarkLinkRewrite` with `buildDynamicMDX`
// because we already use `remarkLinkRewrite` function to remove .mdx? extensions
const clonedRemarkLinkRewrite = remarkLinkRewrite.bind(null)

type CompileMdxOptions = Pick<
  LoaderOptions,
  | 'staticImage'
  | 'search'
  | 'defaultShowCopyCode'
  | 'readingTime'
  | 'latex'
  | 'codeHighlight'
> & {
  mdxOptions?: MdxOptions
  route?: string
  locale?: string
  filePath?: string
  useCachedCompiler?: boolean
  isPageImport?: boolean
  isPageMapImport?: boolean
}

export async function compileMdx(
  source: string,
  {
    staticImage,
    search,
    readingTime,
    latex,
    codeHighlight,
    defaultShowCopyCode,
    route = '',
    locale,
    mdxOptions = {},
    filePath = '',
    useCachedCompiler,
    isPageImport = true,
    isPageMapImport
  }: CompileMdxOptions = {}
) {
  const {
    jsx = false,
    format: _format = 'mdx',
    outputFormat = 'function-body',
    remarkPlugins,
    rehypePlugins,
    rehypePrettyCodeOptions
  }: MdxOptions = mdxOptions

  if (rehypePrettyCodeOptions) {
    throw new Error(
      "`rehypePrettyCodeOptions` is currently unsupported since `rehype-pretty-code` doesn't support `shikiji` package"
    )
  }

  const format =
    _format === 'detect' ? (filePath.endsWith('.mdx') ? 'mdx' : 'md') : _format

  if (isPageMapImport) {
    const compiler = createProcessor({
      format,
      remarkPlugins: [
        remarkFrontmatter, // parse and attach yaml node
        remarkMdxFrontMatter
      ]
    })
    const vFile = await compiler.process(
      filePath ? { value: source, path: filePath } : source
    )
    const content = vFile.toString()

    const index = content.lastIndexOf('function _createMdxContent(props) {')
    const result = content.slice(0, index)

    return { result } as any
  }

  let searchIndexKey: string | null = null
  if (ERROR_ROUTES.has(route)) {
    /* skip */
  } else if (typeof search === 'object') {
    if (search.indexKey) {
      searchIndexKey = search.indexKey(filePath, route, locale)
      if (searchIndexKey === '') {
        searchIndexKey = locale || DEFAULT_LOCALE
      }
    } else {
      searchIndexKey = locale || DEFAULT_LOCALE
    }
  } else if (search) {
    searchIndexKey = locale || DEFAULT_LOCALE
  }

  // https://github.com/shuding/nextra/issues/1303
  const isFileOutsideCWD =
    !isPageImport && path.relative(CWD, filePath).startsWith('..')

  if (isFileOutsideCWD) {
    throw new Error(
      `Unexpected import of "${filePath}" that is outside of working directory, use symlinks instead`
    )
  }

  const isRemoteContent = outputFormat === 'function-body'

  const compiler =
    !useCachedCompiler || isRemoteContent
      ? createCompiler()
      : (cachedCompilerForFormat[format] ??= createCompiler())
  const processor = compiler()

  try {
    const vFile = await processor.process(
      filePath ? { value: source, path: filePath } : source
    )

    const { title, hasJsxInH1, readingTime, structurizedData } = vFile.data as {
      readingTime?: ReadingTime
      structurizedData: StructurizedData
      title?: string
    } & Pick<PageOpts, 'hasJsxInH1'>
    // https://github.com/shuding/nextra/issues/1032
    const result = String(vFile).replaceAll('__esModule', '_\\_esModule')

    const frontMatter = (vFile.data.frontMatter || {}) as FrontMatter

    if (frontMatter.mdxOptions) {
      throw new Error('`frontMatter.mdxOptions` is no longer supported')
    }

    return {
      result,
      ...(title && { title }),
      ...(hasJsxInH1 && { hasJsxInH1 }),
      ...(readingTime && { readingTime }),
      ...(searchIndexKey !== null && { searchIndexKey, structurizedData }),
      ...(isRemoteContent && { toc: vFile.data.headings }),
      frontMatter
    }
  } catch (err) {
    console.error(`[nextra] Error compiling ${filePath}.`)
    throw err
  }

  function createCompiler(): Processor {
    return createProcessor({
      jsx,
      format,
      outputFormat,
      providerImportSource: 'nextra/mdx',
      remarkPlugins: [
        ...(remarkPlugins || []),
        remarkMermaid, // should be before remarkRemoveImports because contains `import { Mermaid } from ...`
        [
          remarkNpm2Yarn, // should be before remarkRemoveImports because contains `import { Tabs as $Tabs, Tab as $Tab } from ...`
          {
            packageName: 'nextra/components',
            tabNamesProp: 'items',
            storageKey: 'selectedPackageManager'
          }
        ] satisfies Pluggable,
        isRemoteContent && remarkRemoveImports,
        remarkFrontmatter, // parse and attach yaml node
        [remarkMdxFrontMatter, { isRemoteContent }] satisfies Pluggable,
        remarkGfm,
        format !== 'md' &&
          ([
            remarkMdxDisableExplicitJsx,
            // Replace the <summary> and <details> with customized components
            { whiteList: ['details', 'summary'] }
          ] satisfies Pluggable),
        remarkCustomHeadingId,
        [remarkHeadings, { isRemoteContent }] satisfies Pluggable,
        // structurize should be before `remarkHeadings` because we attach #id attribute to heading node
        search && ([remarkStructurize, search] satisfies Pluggable),
        staticImage && remarkStaticImage,
        readingTime && remarkReadingTime,
        latex && remarkMath,
        // Remove the markdown file extension from links
        [
          clonedRemarkLinkRewrite,
          {
            pattern: MARKDOWN_URL_EXTENSION_REGEX,
            replace: '',
            excludeExternalLinks: true
          }
        ] satisfies Pluggable
      ].filter(truthy),
      rehypePlugins: [
        ...(rehypePlugins || []),
        format === 'md' && [
          // To render <details /> and <summary /> correctly
          rehypeRaw,
          // fix Error: Cannot compile `mdxjsEsm` node for npm2yarn and mermaid
          { passThrough: ['mdxjsEsm', 'mdxJsxFlowElement'] }
        ],
        [parseMeta, { defaultShowCopyCode }],
        ...(codeHighlight === false
          ? []
          : [
              [rehypePrettyCode, DEFAULT_REHYPE_PRETTY_CODE_OPTIONS] as any,
              attachMeta
            ]),
        latex && rehypeKatex
      ].filter(truthy)
    })
  }
}
