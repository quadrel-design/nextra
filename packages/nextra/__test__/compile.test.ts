import { compileMdx } from '../src/server/compile.js'
import { clean } from './test-utils.js'

const mdxOptions = {
  jsx: true,
  outputFormat: 'program'
} as const

describe('Process heading', () => {
  it('code-h1', async () => {
    const result = await compileMdx('# `codegen.yml`', { mdxOptions })
    expect(result).toMatchSnapshot()
  })
  it('code-with-text-h1', async () => {
    const result = await compileMdx('# `codegen.yml` file', { mdxOptions })
    expect(result).toMatchSnapshot()
  })
  it('static-h1', async () => {
    const result = await compileMdx('# Hello World', { mdxOptions })
    expect(result).toMatchSnapshot()
  })
  it('dynamic-h1', async () => {
    const result = await compileMdx(
      `
import { useRouter } from 'next/router'

export const TagName = () => {
  const { tag } = useRouter().query
  return tag || null
}

# Posts Tagged with ‚Äú<TagName/>‚Äù
    `,
      { mdxOptions }
    )
    expect(result).toMatchSnapshot()
  })
  it('no-h1', async () => {
    const result = await compileMdx('## H2', { mdxOptions })
    expect(result).toMatchSnapshot()
  })
  it('use custom heading id', async () => {
    const { result } = await compileMdx(
      `
# My Header [#test-id]
## Some extra space [#extra-space]&nbsp;
### Some extra space in heading    [#extra-space-in-heading]
### nospace[#without-space]
#### foo [#–¥—Ä—É–≥–æ–π —è–∑—ã–∫]
##### bar Baz []
###### bar Qux [#]`,
      { mdxOptions }
    )
    expect(await clean(result)).toMatchInlineSnapshot(`
      "import { useMDXComponents as _provideComponents } from \\"nextra/mdx\\";
      export const frontMatter = {};
      export const toc = [
        {
          depth: 2,
          value: \\"Some extra space\\",
          id: \\"extra-space\\",
        },
        {
          depth: 3,
          value: \\"Some extra space in heading\\",
          id: \\"extra-space-in-heading\\",
        },
        {
          depth: 3,
          value: \\"nospace\\",
          id: \\"without-space\\",
        },
        {
          depth: 4,
          value: \\"foo\\",
          id: \\"–¥—Ä—É–≥–æ–π-—è–∑—ã–∫\\",
        },
        {
          depth: 5,
          value: \\"bar Baz []\\",
          id: \\"bar-baz-\\",
        },
        {
          depth: 6,
          value: \\"bar Qux [#]\\",
          id: \\"bar-qux-\\",
        },
      ];
      function _createMdxContent(props) {
        const _components = Object.assign(
          {
            h1: \\"h1\\",
            h2: \\"h2\\",
            h3: \\"h3\\",
            h4: \\"h4\\",
            h5: \\"h5\\",
            h6: \\"h6\\",
          },
          _provideComponents(),
          props.components,
        );
        return (
          <>
            <_components.h1 id=\\"test-id\\">{\\"My Header\\"}</_components.h1>
            {\\"\\\\n\\"}
            <_components.h2 id={toc[0].id}>{toc[0].value}</_components.h2>
            {\\"\\\\n\\"}
            <_components.h3 id={toc[1].id}>{toc[1].value}</_components.h3>
            {\\"\\\\n\\"}
            <_components.h3 id={toc[2].id}>{toc[2].value}</_components.h3>
            {\\"\\\\n\\"}
            <_components.h4 id={toc[3].id}>{toc[3].value}</_components.h4>
            {\\"\\\\n\\"}
            <_components.h5 id={toc[4].id}>{toc[4].value}</_components.h5>
            {\\"\\\\n\\"}
            <_components.h6 id={toc[5].id}>{toc[5].value}</_components.h6>
          </>
        );
      }
      "
    `)
  })
  it('use github-slugger', async () => {
    const { result } = await compileMdx('### My Header', { mdxOptions })
    expect(result).toMatch('<_components.h3 id="my-header">{"My Header"}')
  })

  it('should merge headings from partial components', async () => {
    const { result } = await compileMdx(
      `
import FromMdx from './one.mdx'
import FromMarkdown from './two.md'
import IgnoreMe from './foo'

## ‚ù§Ô∏è

<FromMdx />

## ‚úÖ

<FromMarkdown />

import Last from './three.mdx'

<Last />

<IgnoreMe />

## üëã

## kek <Kek />

## \`try\` me

## latex $l$

## {'interpolate'} {1} {true} {null} {variable}

`,
      { mdxOptions, latex: true }
    )
    expect(await clean(result)).toMatchInlineSnapshot(`
      "import { useMDXComponents as _provideComponents } from \\"nextra/mdx\\";
      export const frontMatter = {};
      import FromMdx, { toc as toc0 } from \\"./one.mdx\\";
      import FromMarkdown, { toc as toc1 } from \\"./two.md\\";
      import IgnoreMe from \\"./foo\\";
      import Last, { toc as toc2 } from \\"./three.mdx\\";
      export const toc = [
        {
          depth: 2,
          value: \\"‚ù§Ô∏è\\",
          id: \\"Ô∏è\\",
        },
        ...toc0,
        {
          depth: 2,
          value: \\"‚úÖ\\",
          id: \\"\\",
        },
        ...toc1,
        ...toc2,
        {
          depth: 2,
          value: \\"üëã\\",
          id: \\"-1\\",
        },
        {
          depth: 2,
          value: \\"kek \\",
          id: \\"kek-\\",
        },
        {
          depth: 2,
          value: \\"try me\\",
          id: \\"try-me\\",
        },
        {
          depth: 2,
          value: \\"latex l\\",
          id: \\"latex-l\\",
        },
        {
          depth: 2,
          value: (
            <>
              {\\"interpolate\\"} {1} {true} {null} {variable}
            </>
          ),
          id: \\"interpolate-1-true-null-variable\\",
        },
      ];
      function _createMdxContent(props) {
        const _components = Object.assign(
            {
              h2: \\"h2\\",
              code: \\"code\\",
              span: \\"span\\",
              math: \\"math\\",
              semantics: \\"semantics\\",
              mrow: \\"mrow\\",
              mi: \\"mi\\",
              annotation: \\"annotation\\",
            },
            _provideComponents(),
            props.components,
          ),
          { Kek } = _components;
        if (!Kek) _missingMdxReference(\\"Kek\\", true);
        return (
          <>
            <_components.h2 id={toc[0].id}>{toc[0].value}</_components.h2>
            {\\"\\\\n\\"}
            <FromMdx />
            {\\"\\\\n\\"}
            <_components.h2 id={toc[2].id}>{toc[2].value}</_components.h2>
            {\\"\\\\n\\"}
            <FromMarkdown />
            {\\"\\\\n\\"}
            {\\"\\\\n\\"}
            <Last />
            {\\"\\\\n\\"}
            <IgnoreMe />
            {\\"\\\\n\\"}
            <_components.h2 id={toc[5].id}>{toc[5].value}</_components.h2>
            {\\"\\\\n\\"}
            <_components.h2 id={toc[6].id}>
              {\\"kek \\"}
              <Kek />
            </_components.h2>
            {\\"\\\\n\\"}
            <_components.h2 id={toc[7].id}>
              <_components.code>{\\"try\\"}</_components.code>
              {\\" me\\"}
            </_components.h2>
            {\\"\\\\n\\"}
            <_components.h2 id={toc[8].id}>
              {\\"latex \\"}
              <_components.span className=\\"katex\\">
                <_components.span className=\\"katex-mathml\\">
                  <_components.math xmlns=\\"http://www.w3.org/1998/Math/MathML\\">
                    <_components.semantics>
                      <_components.mrow>
                        <_components.mi>{\\"l\\"}</_components.mi>
                      </_components.mrow>
                      <_components.annotation encoding=\\"application/x-tex\\">
                        {\\"l\\"}
                      </_components.annotation>
                    </_components.semantics>
                  </_components.math>
                </_components.span>
                <_components.span className=\\"katex-html\\" aria-hidden=\\"true\\">
                  <_components.span className=\\"base\\">
                    <_components.span
                      className=\\"strut\\"
                      style={{
                        height: \\"0.6944em\\",
                      }}
                    />
                    <_components.span
                      className=\\"mord mathnormal\\"
                      style={{
                        marginRight: \\"0.01968em\\",
                      }}
                    >
                      {\\"l\\"}
                    </_components.span>
                  </_components.span>
                </_components.span>
              </_components.span>
            </_components.h2>
            {\\"\\\\n\\"}
            <_components.h2 id={toc[9].id}>{toc[9].value}</_components.h2>
          </>
        );
      }
      "
    `)
  })
  it('should not attach headings with parent Tab or Tabs.Tab', async () => {
    const { result } = await compileMdx(
      `
<Tab>
 ## foo
</Tab>

<Tabs.Tab>
  ## bar
  ## baz [#custom-id]
</Tabs.Tab>
`,
      { mdxOptions }
    )
    expect(result).toMatch('export const toc = [];')
    expect(result).not.toMatch('id="custom-id"')
  })
})

describe('Link', () => {
  it('supports .md links', async () => {
    const { result } = await compileMdx('[link](../file.md)', { mdxOptions })
    expect(result).toMatch('<_components.a href="../file">')
  })

  it('supports .mdx links', async () => {
    const { result } = await compileMdx('[link](../file.mdx)', { mdxOptions })
    expect(result).toMatch('<_components.a href="../file">')
  })

  it('supports URL links', async () => {
    const { result } = await compileMdx('[link](../file)', { mdxOptions })
    expect(result).toMatch('<_components.a href="../file">')
  })

  it('supports query', async () => {
    const { result } = await compileMdx('[link](../file.md?query=a)', {
      mdxOptions
    })
    expect(result).toMatch('<_components.a href="../file?query=a">')
  })

  it('supports anchor', async () => {
    const { result } = await compileMdx('[link](../file.md#anchor)', {
      mdxOptions
    })
    expect(result).toMatch('<_components.a href="../file#anchor">')
  })

  it('supports external .md links', async () => {
    const { result } = await compileMdx('[link](https://example.com/file.md)', {
      mdxOptions
    })
    expect(result).toMatch('<_components.a href="https://example.com/file.md">')
  })

  it('supports external .mdx links', async () => {
    const { result } = await compileMdx(
      '[link](https://example.com/file.mdx)',
      {
        mdxOptions
      }
    )
    expect(result).toMatch(
      '<_components.a href="https://example.com/file.mdx">'
    )
  })
})

describe('Code block', () => {
  describe('Filename', () => {
    it('attach with "codeHighlight: true" by default', async () => {
      const { result } = await compileMdx('```text filename="test.js"\n```', {
        mdxOptions
      })
      expect(result).toMatch(
        '<_components.pre data-language="text" data-filename="test.js">'
      )
    })

    it('attach with "codeHighlight: false"', async () => {
      const { result } = await compileMdx('```js filename="test.js"\n```', {
        mdxOptions,
        codeHighlight: false
      })
      expect(result).toMatch('<_components.pre data-filename="test.js">')
    })

    it('not highlight filename as substring', async () => {
      const { result } = await compileMdx('```js filename="/foo/"\nfoo\n```', {
        mdxOptions,
        codeHighlight: true // processed only by rehype-pretty-code
      })
      expect(result).not.toMatch(
        'className="highlighted">{"foo"}</_components.span>'
      )
      expect(result).toMatch('}}>{"foo"}</_components.span>')
    })
  })

  describe('Highlight', () => {
    it('should support line highlights', async () => {
      const { result } = await compileMdx(
        '```js filename="test.js" {1}\n123\n```',
        {
          mdxOptions
        }
      )
      expect(result).toMatch('<_components.span data-highlighted-line="">')
    })
  })

  describe('Copy code button', () => {
    for (const codeHighlight of [true, false]) {
      describe(`codeHighlight: ${codeHighlight}`, () => {
        it('attach with "copy"', async () => {
          const { result } = await compileMdx('```js copy\n```', {
            mdxOptions,
            codeHighlight
          })
          expect(result).toMatch('data-copy="">')
        })

        it('attach with "defaultShowCopyCode: true"', async () => {
          const { result } = await compileMdx('```js\n```', {
            mdxOptions,
            defaultShowCopyCode: true,
            codeHighlight
          })
          expect(result).toMatch('data-copy="">')
        })

        it('not attach with "defaultShowCopyCode: true" and "copy=false"', async () => {
          const { result } = await compileMdx('```js copy=false\n```', {
            mdxOptions,
            defaultShowCopyCode: true,
            codeHighlight
          })
          expect(result).not.toMatch('data-copy="">')
        })
      })
    }
  })
})
