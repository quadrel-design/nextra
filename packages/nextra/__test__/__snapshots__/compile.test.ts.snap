// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Process heading > code-h1 1`] = `
"import { useMDXComponents as _provideComponents } from \\"nextra/mdx\\";
export const frontMatter = {};
export function useTOC(props) {
  return [];
}
function _createMdxContent(props) {
  const { toc } = props;
  const _components = Object.assign(
    {
      h1: \\"h1\\",
      code: \\"code\\",
    },
    _provideComponents(),
    props.components,
  );
  return (
    <_components.h1>
      <_components.code>{\\"codegen.yml\\"}</_components.code>
    </_components.h1>
  );
}
"
`;

exports[`Process heading > code-with-text-h1 1`] = `
"import { useMDXComponents as _provideComponents } from \\"nextra/mdx\\";
export const frontMatter = {};
export function useTOC(props) {
  return [];
}
function _createMdxContent(props) {
  const { toc } = props;
  const _components = Object.assign(
    {
      h1: \\"h1\\",
      code: \\"code\\",
    },
    _provideComponents(),
    props.components,
  );
  return (
    <_components.h1>
      <_components.code>{\\"codegen.yml\\"}</_components.code>
      {\\" file\\"}
    </_components.h1>
  );
}
"
`;

exports[`Process heading > dynamic-h1 1`] = `
{
  "frontMatter": {},
  "hasJsxInH1": true,
  "result": "import { useMDXComponents as _provideComponents } from \\"nextra/mdx\\";
export const frontMatter = {};
import { useRouter } from \\"next/router\\";
export const TagName = () => {
  const { tag } = useRouter().query;
  return tag || null;
};
export function useTOC(props) {
  return [];
}
function _createMdxContent(props) {
  const { toc } = props;
  const _components = Object.assign(
    {
      h1: \\"h1\\",
    },
    _provideComponents(),
    props.components,
  );
  return (
    <_components.h1>
      {\\"Posts Tagged with “\\"}
      <TagName />
      {\\"”\\"}
    </_components.h1>
  );
}
",
  "title": "Posts Tagged with “”",
}
`;

exports[`Process heading > no-h1 1`] = `
"import { useMDXComponents as _provideComponents } from \\"nextra/mdx\\";
export const frontMatter = {};
export function useTOC(props) {
  return [
    {
      value: \\"H2\\",
      id: \\"h2\\",
      depth: 2,
    },
  ];
}
function _createMdxContent(props) {
  const { toc } = props;
  const _components = Object.assign(
    {
      h2: \\"h2\\",
    },
    _provideComponents(),
    props.components,
  );
  return <_components.h2 id={toc[0].id}>{toc[0].value}</_components.h2>;
}
"
`;

exports[`Process heading > static-h1 1`] = `
"import { useMDXComponents as _provideComponents } from \\"nextra/mdx\\";
export const frontMatter = {};
export function useTOC(props) {
  return [];
}
function _createMdxContent(props) {
  const { toc } = props;
  const _components = Object.assign(
    {
      h1: \\"h1\\",
    },
    _provideComponents(),
    props.components,
  );
  return <_components.h1>{\\"Hello World\\"}</_components.h1>;
}
"
`;
